File: server_python/database/anime.py
from database.abstract import AbstractDataObject
from shared.config import SONG_LIST_PATH
from shared.exceptions import AMQDbError

class SongData(AbstractDataObject):

    def __init__(self, logger):
        super().__init__(SONG_LIST_PATH, logger)
        self._animeChoices = []
        self._songChoices = []
        self._songIds = []
        self._initChoices()

    def _initDb(self):
        self.db = self._readFile()
        self.db.sort(key=lambda x: (x.get('anime', [''])[0].lower(), x.get('title').lower()))

    def _initChoices(self):
        self._animeChoices = []
        self._songChoices = []
        self._songIds = []

        for song in self.db:
            self._addSongId(song)
            self._addSongChoice(song)
            self._addAnimeChoice(song)

        self._animeChoices.sort()
        self._songChoices.sort()

    @property
    def choices(self):
        return {
            'anime': self._animeChoices,
            'song': self._songChoices
        }

    def validateSongId(self, user, songId):
        if songId not in self._songIds:
            self._logger.writeLog('DATA001', { 'user': user,
                                               'songId': songId,
                                               'reason': 'invalid song id' })
            raise AMQDbError('Invalid song ID')

    def _addAnimeChoice(self, song):
        for anime in song['anime']:
            if anime not in self._animeChoices:
                self._animeChoices.append(anime)

    def _addSongChoice(self, song):
        title = song['title']
        if title not in self._songChoices:
            self._songChoices.append(title)

    def _addSongId(self, song):
        songId = song['songId']
        if songId in self._songIds:
            raise Exception(f'Duplicate song id: {songId}')
        else:
            self._songIds.append(songId)

File: server_ts/src/database/anime.ts
import { AbstractDataObject } from './abstract'
import { SONG_LIST_PATH } from '../shared/config'
import { AMQLogger } from '../shared/logging/logger'
import { Song } from '../shared/interfaces'
import { AMQDbError } from '../shared/exceptions'

class SongData extends AbstractDataObject {
  db: Array<Song>
  _songChoices: Array<string>
  _animeChoices: Array<string>
  _songIds: Array<string>

  constructor(logger: AMQLogger) {
    super(SONG_LIST_PATH, logger)
    this._initSecondaryDb()
  }

  validateSongId(user: string, songId: string): void {
    if (!this._songIds.includes(songId)) {
      this._logger.writeLog('DATA001', { user: user,
                                         songId: songId,
                                         reason: 'invalid song id' })
      throw new AMQDbError('Invalid song ID')
    }
  }

  _initSecondaryDb(): void {
    this._animeChoices = []
    this._songChoices = []
    this._songIds = []

    for (let song of this.db) {
      this._addSongId(song)
      this._addAnimeChoice(song)
    }

    this._animeChoices = this._animeChoices.sort(this._sortChoices)
    this._songChoices = this._songChoices.sort(this._sortChoices)
  }

  _addSongChoice(song: Song): void {
    let title = song['title']
    if (!this._songChoices.includes(title)) {
      this._songChoices.push(title)
    }
  }

  _addAnimeChoice(song: Song): void {
    for (let anime of song['anime']) {
      if (!this._animeChoices.includes(anime)) {
        this._animeChoices.push(anime)
      }
    }
  }

  _addSongId(song: Song): void {
    let songId = song['songId']
    if (this._songIds.includes(songId)) {
      throw new Error(`Duplicate song id: ${songId}`)
    }
    else {
      this._songIds.push(songId)
    }
  }

  _sortChoices(a: string, b: string): number {
    if (a > b) {
      return 1
    }
    return -1
  }
}

export { SongData }